### PR TITLE
Starlight collator rotation

### PR DESCRIPTION
**Pull request by fgamundi**
Enables collator rotation in Starlight

### CODE REVIEW

**Comment by fgamundi** [comment]
@girazoki I've added a TS test. Let me know if I can improve it

File: solo-chains/runtime/starlight/src/lib.rs [hunk 1]
```diff
@@ -2747,6 +2749,85 @@ impl tanssi_initializer::Config for Runtime {
...
+            } {
+                // Return random_hash as a [u8; 32] instead of a Hash
+                let mut buf = [0u8; 32];
+                let len = sp_std::cmp::min(32, random_hash.as_ref().len());
+                buf[..len].copy_from_slice(&random_hash.as_ref()[..len]);
+
+                buf
+            } else {
+                // If there is no randomness (e.g when running in dev mode), return [0; 32]
+                // TODO: smoke test to ensure this never happens in a live network
```
Comment (tmpolaczyk) [inline solo-chains/runtime/starlight/src/lib.rs:2809]: Btw this test already exists `test-randomness-consistency.ts`, you can remove this comment from all the runtimes

```suggestion
```

File: solo-chains/runtime/starlight/src/lib.rs [hunk 2]
```diff
@@ -2747,6 +2749,85 @@ impl tanssi_initializer::Config for Runtime {
...
+                BabeCurrentBlockRandomnessGetter::get_block_randomness_mixed(b"CollatorAssignment")
+            } {
+                // Return random_hash as a [u8; 32] instead of a Hash
+                let mut buf = [0u8; 32];
+                let len = sp_std::cmp::min(32, random_hash.as_ref().len());
+                buf[..len].copy_from_slice(&random_hash.as_ref()[..len]);
+
+                buf
+            } else {
+                // If there is no randomness (e.g when running in dev mode), return [0; 32]
```
Comment (tmpolaczyk) [inline solo-chains/runtime/starlight/src/lib.rs:2808]: So in starlight dev mode there is randomness? Or did you have to do something in the test to add randomness?

File: solo-chains/runtime/starlight/src/lib.rs [hunk 3]
```diff
@@ -2747,6 +2749,85 @@ impl tanssi_initializer::Config for Runtime {
...
+                // TODO: smoke test to ensure this never happens in a live network
+                [0; 32]
+            }
+        } else {
+            // In block 0 (genesis) there is randomness
```
Comment (tmpolaczyk) [inline solo-chains/runtime/starlight/src/lib.rs:2813]: This typo is probably mine

```suggestion
            // In block 0 (genesis) there is no randomness
```

File: solo-chains/runtime/starlight/src/lib.rs (hunk 2 above)
Comment (girazoki) [reply solo-chains/runtime/starlight/src/lib.rs:2808]: there was randomness yes

File: solo-chains/runtime/starlight/src/lib.rs [hunk 4]
```diff
@@ -2747,6 +2749,85 @@ impl tanssi_initializer::Config for Runtime {
     type SessionHandler = OwnApplySession;
 }
 
+pub struct BabeCurrentBlockRandomnessGetter;
+impl BabeCurrentBlockRandomnessGetter {
+    fn get_block_randomness() -> Option<[u8; 32]> {
+        Babe::author_vrf_randomness()
```
Comment (girazoki) [inline solo-chains/runtime/starlight/src/lib.rs:2755]: comment

Comment (girazoki) [reply solo-chains/runtime/starlight/src/lib.rs:2755]: specifically that we use the previous block randomness

File: solo-chains/runtime/starlight/src/lib.rs [hunk 5]
```diff
@@ -2747,6 +2749,85 @@ impl tanssi_initializer::Config for Runtime {
...
+impl Get<u32> for ConfigurationCollatorRotationSessionPeriod {
+    fn get() -> u32 {
+        CollatorConfiguration::config().full_rotation_period
+    }
+}
+
+pub struct BabeGetRandomnessForNextBlock;
+
+impl GetRandomnessForNextBlock<u32> for BabeGetRandomnessForNextBlock {
+    fn should_end_session(n: u32) -> bool {
```
Comment (girazoki) [inline solo-chains/runtime/starlight/src/lib.rs:2786]: Same thing, comment. Why are we doing this? (I know we discussed it the other day, but it is not trivial for a new comer)

File: solo-chains/runtime/starlight/src/lib.rs (hunk 2 above)
Comment (girazoki) [reply solo-chains/runtime/starlight/src/lib.rs:2808]: this comment has to be removed acutally (the one referring dev mode)

File: solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs [hunk 6]
```diff
@@ -31,6 +30,63 @@ use {
...
+            assert!(TanssiCollatorAssignment::pending_collator_container_chain().is_none());
+
+            run_to_session(rotation_period - 1);
+            assert_eq!(
+                TanssiCollatorAssignment::collator_container_chain(),
+                initial_assignment,
+            );
+            assert!(TanssiCollatorAssignment::pending_collator_container_chain().is_some());
+
+            run_to_session(rotation_period);
```
Comment (girazoki) [inline solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs:81]: Ideally we want to test that before the new session block the randomness is populated into COllatorAssignment here. I hope btw this is deterministic

File: solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs [hunk 7]
```diff
@@ -1939,3 +1995,106 @@ fn test_collator_assignment_tip_withdraw_min_tip() {
...
+            // Alice gets 10k extra tokens for her mapping deposit
+            (AccountId::from(ALICE), 210_000 * UNIT),
+            (AccountId::from(BOB), 100_000 * UNIT),
+            (AccountId::from(CHARLIE), 100_000 * UNIT),
+            (AccountId::from(DAVE), 100_000 * UNIT),
+        ])
+        .with_collators(vec![
+            (AccountId::from(ALICE), 210 * UNIT),
+            (AccountId::from(BOB), 100 * UNIT),
+            // (AccountId::from(CHARLIE), 100 * UNIT),
```
Comment (girazoki) [inline solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs:2012]: remove

File: solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs [hunk 8]
```diff
@@ -1939,3 +1995,106 @@ fn test_collator_assignment_tip_withdraw_min_tip() {
...
+            let assignment = TanssiCollatorAssignment::collator_container_chain();
+            assert_eq!(
+                assignment.container_chains[&1002u32.into()],
+                vec![ALICE.into(), BOB.into()]
+            );
+        });
+}
+
+#[test]
+fn test_parachains_deregister_collators_config_change_reassigned() {
```
Comment (girazoki) [inline solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs:2051]:
```suggestion
fn test_parachains_collators_config_change_reassigned() {
```

File: test/suites/dev-tanssi-relay/collator-assignment/test-collator-assignment.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                );
+
+                // remainingSessionsForRotation
+                await jumpSessions(context, 1);
+
+                const finalAssignment = (
+                    await polkadotJs.query.tanssiCollatorAssignment.collatorContainerChain()
+                ).toJSON();
+                // Assignment should have changed
+                expect(initialAssignment.containerChains[2000].toSorted()).to.not.deep.eq(
```
Comment (girazoki) [inline test/suites/dev-tanssi-relay/collator-assignment/test-collator-assignment.ts:57]: Same question applies. is this deterministic? can you verify that the author-vrf is always the same when you run it multiple times?

File: solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs [hunk 10]
```diff
@@ -31,6 +30,63 @@ use {
...
+        .with_collators(vec![
+            (AccountId::from(ALICE), 210 * UNIT),
+            (AccountId::from(BOB), 100 * UNIT),
+            (AccountId::from(CHARLIE), 100 * UNIT),
+            (AccountId::from(DAVE), 100 * UNIT),
+        ])
+        .with_empty_parachains(vec![1001, 1002])
+        .with_config(pallet_configuration::HostConfiguration {
+            max_collators: 100,
+            min_orchestrator_collators: 2,
```
Comment (girazoki) [inline solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs:52]: this numbers can be 0 I guess, it does not make sense to have orchestrator collators in staright

File: solo-chains/runtime/starlight/src/lib.rs (hunk 1 above)
Comment (fgamundi) [reply solo-chains/runtime/starlight/src/lib.rs]: fixed

File: solo-chains/runtime/starlight/src/lib.rs (hunk 2 above)
Comment (fgamundi) [reply solo-chains/runtime/starlight/src/lib.rs]: fixed

File: solo-chains/runtime/starlight/src/lib.rs (hunk 3 above)
Comment (fgamundi) [reply solo-chains/runtime/starlight/src/lib.rs]: fixed

File: solo-chains/runtime/starlight/src/lib.rs (hunk 4 above)
Comment (fgamundi) [reply solo-chains/runtime/starlight/src/lib.rs:2756]: Comment added, let me know if it's clear

File: solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs (hunk 6 above)
Comment (fgamundi) [reply solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs]: Updated both integration and TS tests

File: solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs (hunk 7 above)
Comment (fgamundi) [reply solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs]: fixed

File: solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs (hunk 8 above)
Comment (fgamundi) [reply solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs]: fixed

File: test/suites/dev-tanssi-relay/collator-assignment/test-collator-assignment.ts (hunk 9 above)
Comment (fgamundi) [reply test/suites/dev-tanssi-relay/collator-assignment/test-collator-assignment.ts]: Confirmed that it's not deterministic in dev mode. Added a check to the tests that the randomness seed in `CollatorAssignment` is set before the full rotation

File: solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs (hunk 10 above)
Comment (fgamundi) [reply solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs]: fixed

**Review by girazoki — APPROVED** [review]
LGTM

File: solo-chains/runtime/starlight/src/lib.rs (hunk 4 above)
Comment (girazoki) [reply solo-chains/runtime/starlight/src/lib.rs:2756]: resolved

File: solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs (hunk 6 above)
Comment (girazoki) [reply solo-chains/runtime/starlight/src/tests/collator_assignment_tests.rs]: resolved

File: test/suites/dev-tanssi-relay/collator-assignment/test-collator-assignment.ts (hunk 9 above)
Comment (girazoki) [reply test/suites/dev-tanssi-relay/collator-assignment/test-collator-assignment.ts]: resolved

File: solo-chains/runtime/starlight/src/lib.rs (hunk 5 above)
Comment (girazoki) [reply solo-chains/runtime/starlight/src/lib.rs:2789]: resolved

**Review by tmpolaczyk — APPROVED** [review]

**Pull request merged**